packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts

```diff
@@ -0,0 +1,149 @@
+import { Resource, TagType } from '../cloudformation/resource';
+import { Construct, IAspect } from '../core/construct';
+
+export interface TagAspectProps extends TagProperties {
+  // TODO docs
+  value?: string;
+}
+
+export interface TagProperties {
+  // TODO docs
+  propagate?: number;
+  include?: string[];
+  exclude?: string[];
+}
+
+export abstract class TagBaseAspect implements IAspect {
+
+  public readonly type: string = 'taggable';
+
+  public readonly key: string;
+  public readonly value?: string;
+
+  public propagate: number;
+  private readonly include: string[];
+  private readonly exclude: string[];
+  private readonly visitedBy: {[id: string]: boolean} = {};
+
+  private parent?: Construct;
+
+  constructor(key: string, props: TagAspectProps = {}) {
+    this.key = key;
+
+    this.propagate = props.propagate === undefined ? -1 : props.propagate;
+    this.value = props.value;
+    this.include = props.include || [];
+    this.exclude = props.exclude || [];
+  }
+
+  public visit(node: Construct): void {
+    if (this.visitedBy[node.uniqueId] === true) {
+      return;
+    }
+    this.visitedBy[node.uniqueId] = true;
+    const propagate = this.propagate > 0 ? this.propagate - 1 : this.propagate;
+    this.parent = this.parent || node;
+    if (propagate !== 0) {
+      for (const child of node.children) {
+        this.propagate = propagate;
+        child.apply(this);
+      }
+    }
+    if (node instanceof Resource && (node as Resource).tagType !== TagType.NotTaggable) {
+      const resource = node as Resource;
+      if (this.exclude.length !== 0 && this.exclude.indexOf(resource.resourceType) !== -1) {
+        return;
+      }
+      if (this.include.length !== 0 && this.include.indexOf(resource.resourceType) === -1) {
+        return;
+      }
+      if (this.hasPrecedence(node.aspects)) {
+        this.applyTag(resource);
+      }
+    }
+  }
+
+  protected abstract applyTag(resource: Resource): void;
+
+  private hasPrecedence(aspects: IAspect[]): boolean {
+    const tagAspects = aspects.filter( a => a !== this && a.type === this.type);
+    const collisions = tagAspects.filter( a => (a as TagBaseAspect).key === this.key);
+    for (const collidingAspect of collisions) {
+      const collision = collidingAspect as TagBaseAspect;
+      // if the parent is not defined the apsect hasn't been visited and is
+      // local to the current node; thus the collision has precedence because it
+      // is later (e.g. last one in wins) in the array
+      if (collision.parent === undefined) {
+        return false;
+      }
+      // the aspect with the most ancestors is closest to this node in the tree
+      if (collision.parent.ancestors().length > this.parent!.ancestors().length) {
+        return false;
+      }
+    }
+    return true;
+  }
+}
+
+export class TagAspect extends TagBaseAspect {
```

Comment:
I know this is going to be contentious but I think I'd prefer this to *not* be suffixed with `Aspect`.

```
construct.apply(new cdk.Tag('key', 'value'));
```

Reads better to me.

Reply:
I'll agree this reads better and I originally I started here. For this draft I didn't want to rename the current `cdk.cloudformation.Tag` class that we have. I did some looking and the specific class is mostly used in generated code. Would we prefer that we rename this to `CfnTag`. I realize `Cfn<resource>` now has special meaning and I think this is similar, but not identical. I also felt like the fact `apply` (or another name we choose) will already set the mental context of the developer that this is an `Aspect` and suffix is not necessary.